### ISSUE TITLE
CA1416: False positive when using in a LINQ statement

### ISSUE THREAD

**Issue by AnthonyIacono** [dotnet/roslyn-analyzers#6439]
### Analyzer

**Diagnostic ID**: [CA1416](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1416): `Validate platform compatibility`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 7 SDK](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 7.0.100](https://dotnet.microsoft.com/download/dotnet/7.0)

### Describe the bug

The CA1416 analyzer will incorrectly report a warning that platform-code is reachable when it is in a LINQ statement, even if the execution is not deferred.

### Steps To Reproduce

```
using System.Runtime.Versioning;

public static class Program
{
    public static void Main(string[] args)
    {
        if (OperatingSystem.IsLinux())
        {
            var foo = new Foo();

            // Good
            var works = foo.LinuxOnlyApi();

            // Bad
            var foos = new List<Foo> { foo };
            var broken = foos.Select(f => f.LinuxOnlyApi()).FirstOrDefault();
        }
    }
}

public class Foo
{

    [SupportedOSPlatform("linux")]
    public int LinuxOnlyApi() { return 42; }
}
```

### Expected behavior

Neither instance of `LinuxOnlyApi` should report `CA1416`.

### Actual behavior

The instance of `LinuxOnlyApi` within the `.Select()` of the LINQ statement reports `CA1416`.

**Comment by yaakov-h** [comment]
I have a similar case which I think is the same issue, for a Windows-only API:

```csharp
public void Foo(bool flag, IntPtr affinity)
{
    if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && flag)
    {
        UpdateProcess(p => p.ProcessorAffinity = affinity);
    }
}

static void UpdateProcess(Action<Process> func) => throw null!;
```

> warning CA1416: This call site is reachable on all platforms. 'Process.ProcessorAffinity' is only supported on: 'linux', 'windows'.

**Comment by jnm2** [comment]
I filed the issue as https://github.com/dotnet/roslyn-analyzers/issues/7323 because this is specific to lambdas, not per se LINQ.

**Comment by buyaa-n** [comment]
This might be by design, same as https://github.com/dotnet/roslyn-analyzers/issues/4282

@mavasani is this scenario could be fixed within flow analysis?

**Comment by mavasani** [comment]
@buyaa-n We already handle method calls inside lambdas in flow analysis. This feels like a bug in either the flow analysis component or the platform compat analyzer part that decodes the results and performs analysis.

**Comment by jnm2** [comment]
This one looks very similar to https://github.com/dotnet/roslyn-analyzers/issues/4282, and I commented there proposing something which would resolve the examples in here, in https://github.com/dotnet/roslyn-analyzers/issues/7323, and in the initial example in https://github.com/dotnet/roslyn-analyzers/issues/4282.
